[mlir][spirv] Add folding for Bitwise[Or|And|Xor]

inbelic opened the pull request:
Add missing constant propogation folder for Bitwise[Or|And|Xor].

Move previous Bitwise[Or|And] fold implementations to SPIRVCanonicalization for consistency.

Implement additional folding when lhs == rhs and rhs = 0 for Xor. As well as, update an Xor testcase to account for this introduced folding.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

inbelic committed:
[mlir][spirv] Add folding for Bitwise[Or|And|Xor]

Add missing constant propogation folder for Bitwise[Or|And|Xor].

Move previous Bitwise[Or|And] fold implementations to
SPIRVCanonicalization for consistency.

Implement additional folding when lhs == rhs and rhs = 0 for Xor.
As well as, update an Xor testcase to account for this introduced folding.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

kuhar reviewed (commented):
This looks very useful, thanks for the contribution @inbelic!

I left a few comments, but overall looks good already.

inbelic committed:
review comments:

- add folding for x & x -> x, and, x | x -> x
- extend testcases to display vector capabilities

kuhar reviewed (approved):
LGTM, thanks for the fixes
